torchmetrics/metric.py

```diff
@@ -411,6 +413,101 @@ def __setstate__(self, state: Dict[str, Any]) -> None:
         self.compute: Callable = self._wrap_compute(self.compute)  # type: ignore
 
+    @property
+    def dtype(self):
+        return self._dtype
+
+    @dtype.setter
+    def dtype(self, new_dtype) -> None:
+        # necessary to avoid infinite recursion
+        raise RuntimeError("Cannot set the dtype explicitly. Please use module.to(new_dtype).")
+
+    @property
+    def device(self):
+        return self._device
+
+    def to(self, *args: Any, **kwargs: Any) -> "Metric":
+        """Moves and/or casts the parameters and buffers. """
+        # there is diff nb vars in PT 1.5
+        out = torch._C._nn._parse_to(*args, **kwargs)
+        self._update_properties(device=out[0], dtype=out[1])
+        return super().to(*args, **kwargs)
+
+    def cuda(self, device: Optional[Union[torch.device, int]] = None) -> "Metric":
+        """Moves all model parameters and buffers to the GPU.
+        This also makes associated parameters and buffers different objects. So
+        it should be called before constructing optimizer if the module will
+        live on GPU while being optimized.
+
+        Arguments:
+            device: if specified, all parameters will be
+                copied to that device
+
+        Returns:
+            Module: self
+        """
+        if device is None or isinstance(device, int):
+            device = torch.device("cuda", index=device)
+        self._update_properties(device=device)
+        return super().cuda(device=device)
+
+    def cpu(self) -> "Metric":
+        """Moves all model parameters and buffers to the CPU.
+
+        Returns:
+            Module: self
+        """
+        self._update_properties(device=torch.device("cpu"))
+        return super().cpu()
+
+    def type(self, dst_type: Union[str, torch.dtype]) -> "Metric":
+        """Casts all parameters and buffers to :attr:`dst_type`.
+
+        Arguments:
+            dst_type (type or string): the desired type
+
+        Returns:
+            Module: self
+        """
+        self._update_properties(dtype=dst_type)
+        return super().type(dst_type=dst_type)
+
+    def float(self) -> "Metric":
+        """Casts all floating point parameters and buffers to ``float`` datatype.
+
+        Returns:
+            Module: self
+        """
+        self._update_properties(dtype=torch.float)
+        return super().float()
+
+    def double(self) -> "Metric":
+        """Casts all floating point parameters and buffers to ``double`` datatype.
+
+        Returns:
+            Module: self
+        """
+        self._update_properties(dtype=torch.double)
+        return super().double()
+
+    def half(self) -> "Metric":
+        """Casts all floating point parameters and buffers to ``half`` datatype.
+
+        Returns:
+            Module: self
+        """
+        self._update_properties(dtype=torch.half)
+        return super().half()
+
+    def _update_properties(
```

Comment:
Just to make sure, this is also applied to the states by the overwritten `_apply` below, right?

Reply:
yes, since every call to `super().half()/cpu() ect` looks something like this:
```python
def half(self: T) -> T:
        r"""Casts all floating point parameters and buffers to ``half`` datatype.

        .. note::
            This method modifies the module in-place.

        Returns:
            Module: self
        """
        return self._apply(lambda t: t.half() if t.is_floating_point() else t)
```